node/impl/full/actor_events.go

```diff
@@ -39,188 +40,320 @@ type ActorEventsAPI struct {
+	if evtFilter == nil {
+		evtFilter = &types.ActorEventFilter{}
+	}
+	params, err := a.parseFilter(*evtFilter)
 	if err != nil {
 		return nil, err
 	}
 
-	fm, err := a.EventFilterManager.Install(ctx, params.MinHeight, params.MaxHeight, params.TipSetCid, f.Filter.Addresses, f.Filter.Fields, false)
+	tipSetCid, err := params.GetTipSetCid()
+	if err != nil {
+		return nil, fmt.Errorf("failed to get tipset cid: %w", err)
+	}
+	fm, err := a.EventFilterManager.Install(ctx, params.MinHeight, params.MaxHeight, tipSetCid, evtFilter.Addresses, evtFilter.Fields, false)
 	if err != nil {
 		return nil, err
 	}
 
-	out := make(chan *types.ActorEvent, 25)
+	// The goal for the code below is to be able to send events on the `out` channel as fast as
+	// possible and not let it get too far behind the rate at which the events are generated.
+	// For historical events we see the rate at which they were generated by looking the height range;
+	// we then make sure that the client can receive them at least twice as fast as they were
+	// generated so they catch up quick enough to receive new events.
+	// For ongoing events we use an exponential moving average of the events per height to make sure
+	// that the client doesn't fall behind.
+	// In both cases we allow a little bit of slack but need to avoid letting the client bloat the
+	// buffer too much.
+	// There is no special handling for reverts, so they will just look like a lot more events per
+	// epoch and the user has to receive them anyway.
+
+	out := make(chan *types.ActorEvent)
 
 	go func() {
 		defer func() {
-			// Tell the caller we're done
+			// tell the caller we're done
 			close(out)
-
-			// Unsubscribe.
 			fm.ClearSubChannel()
-			_ = a.EventFilterManager.Remove(ctx, fm.ID())
+			if err := a.EventFilterManager.Remove(ctx, fm.ID()); err != nil {
+				log.Warnf("failed to remove filter: %s", err)
+			}
 		}()
 
-		if f.Prefill {
-			evs, err := getCollected(ctx, fm)
-			if err != nil {
-				log.Errorf("failed to get collected events: %w", err)
-				return
+		// Handle any historical events that our filter may have picked up -----------------------------
+
+		evs, minEpoch, maxEpoch := getCollected(ctx, fm)
+		if len(evs) > 0 {
+			// must be able to send events at least twice as fast as they were generated
+			epochRange := maxEpoch - minEpoch
+			if epochRange <= 0 {
+				epochRange = 1
+			}
+			eventsPerEpoch := float64(len(evs)) / float64(epochRange)
+			eventsPerSecond := 2 * eventsPerEpoch / float64(build.BlockDelaySecs)
+			// a minimum rate of 1 event per second if we don't have many events
+			if eventsPerSecond < 1 {
+				eventsPerSecond = 1
 			}
 
+			// send events from evs to the out channel and ensure we don't do it slower than eventsPerMs
+			ticker := time.NewTicker(time.Second)
+			defer ticker.Stop()
+
+			const maxSlowTicks = 3 // slightly forgiving, allow 3 slow ticks (seconds) before giving up
+			slowTicks := 0
+			sentEvents := 0.0
```

Comment:
Given that we're doing a single batch here, there's no need for a ticker. We can just have a global timeout.

Reply:
(I mean, the _right_ way to handle slow reads is in the JSON-RPC library... anyways)